pyiron_base/project/update/pyiron_base_03x_to_04x.py

```diff
@@ -0,0 +1,89 @@
+"""
+pyiron_base<=0.3.10 has a bug that writes all arrays with dtype=object even
+numeric ones.  As a fix pyiron_base=0.4.1 introduces a conversion when reading
+such arrays, but does not automatically save them.  This conversion script
+simply goes over all jobs and rewrites their HDF5 files, since it's read with
+the correct dtype, this then writes this correct dtype.
+"""
+
+import os
+import re
+import stat
+import subprocess
+
+from tqdm.auto import tqdm
+
+
+def detect_bug(file_name):
+    """
+    Checks whether HDF5 file has at least one group setup like
+    /foo       Group
+    /foo/data  Dataset {...}
+    /foo/index Dataset {...}
+    which is how h5io stores dtype=object arrays.  If a file doesn't have any
+    of them there's no need to rewrite them.  If there is it might be a
+    corrupted record from our bug or a legitimate dtype=object array.  In that
+    case just rewrite anyway.
+    """
+    out = subprocess.getoutput(f"h5ls -r {file_name}")
+    lines = out.split("\n")
+    for i, l in enumerate(lines[:-2]):
+        if not l.endswith("Group"):
+            continue
+        group_name = l.split()[0]
+        data_match = re.match(
+            f"^{group_name}/data[ \t]*Dataset {'{.*}'}$", lines[i + 1]
+        )
+        index_match = re.match(
+            f"^{group_name}/index[ \t]*Dataset {'{.*}'}$", lines[i + 2]
+        )
+        if data_match and index_match:
+            return True
+    return False
+
+
+def pyiron_base_03x_to_04x(project):
+    """
+    pyiron_base<=0.3.10 has a bug that writes all arrays with dtype=object even
+    numeric ones.  As a fix pyiron_base=0.4.1 introduces a conversion when reading
+    such arrays, but does not automatically save them.  This conversion script
+    simply goes over all jobs and rewrites their HDF5 files, since it's read with
+    the correct dtype, this then writes this correct dtype.
+    """
+    total_size = 0
+    for l in subprocess.getoutput(
+        f"find {project.path} -regex \".*\.h5\" -exec wc -c '{{}}' \;"
```

Comment:
This was my quick hack to get a progress bar working, but it feels a bit hacky now.  I suppose it would work on all linux distros, but might break on windows/mac.

OTOH, this really only counts the number of bytes in all HDF5 files, so maybe you can whip something up with `pyfileindex`/`os.walk` and `os.stat`.

Reply:
I will think about it. Getting rid of the `subprocess` call would be an added benefit.